File: qa/src/report.mjs
import { writeFile, mkdir } from 'node:fs/promises';
import path from 'node:path';
import { rank, LABEL } from './severity.mjs';
import { summariseRegressions } from './checks/regressions.mjs';

const ORDER = ['FAIL', 'WARN', 'INFO', 'PASS'];

/**
 * A site-wide defect in a theme template produces one identical finding per
 * page. Reported raw, a single misconfigured viewport tag becomes 206 failures
 * and buries everything else. Collapse identical findings into one, carrying
 * the page count and a sample of URLs.
 *
 * Grouping is by check + severity + title + a URL-stripped detail signature, so
 * only genuinely repeated findings merge; anything with page-specific substance
 * in its detail stays separate.
 */
export function collapseFindings(findings, { threshold = 3, sampleUrls = 5 } = {}) {
  const groups = new Map();
  const order = [];

  for (const f of findings) {
    const key = `${f.severity}|${f.title}|${signature(f.detail)}`;
    if (!groups.has(key)) {
      groups.set(key, { first: f, items: [] });
      order.push(key);
    }
    groups.get(key).items.push(f);
  }

  const out = [];
  for (const key of order) {
    const { first, items } = groups.get(key);
    if (items.length < threshold) {
      out.push(...items);
      continue;
    }
    const urls = [...new Set(items.map((i) => i.url).filter(Boolean))];
    out.push({
      ...first,
      title: `${first.title} — on ${urls.length || items.length} page(s)`,
      detail: [
        first.detail,
        `Affects ${urls.length || items.length} page(s), e.g.:`,
        ...urls.slice(0, sampleUrls).map((u) => `  ${u}`),
        urls.length > sampleUrls ? `  ...and ${urls.length - sampleUrls} more` : null,
      ].filter(Boolean).join('\n'),
      occurrences: items.length,
      urls,
    });
  }
  return out;
}

/** Detail text with site URLs and digits removed, so per-page noise groups. */
function signature(detail) {
  return String(detail ?? '')
    .replace(/https?:\/\/\S+/g, '<url>')
    .replace(/\d+/g, '<n>')
    .slice(0, 400);
}

export function buildReport({ results, site, cfg, rendered, regressionOutcomes, meta }) {
  const collapse = cfg.report?.collapseRepeatedFindings !== false;
  if (collapse) {
    results = results.map((r) => ({
      ...r,
      rawFindingCount: r.findings.length,
      findings: collapseFindings(r.findings, { threshold: cfg.report?.collapseThreshold ?? 3 }),
    }));
  }

  const counts = { FAIL: 0, WARN: 0, INFO: 0, PASS: 0 };
  for (const r of results) {
    for (const f of r.findings) counts[f.severity] = (counts[f.severity] || 0) + 1;
  }

  const overall = counts.FAIL > 0 ? 'FAIL' : counts.WARN > 0 ? 'WARNING' : 'PASS';
  const regressions = regressionOutcomes ? summariseRegressions(regressionOutcomes) : [];

  return {
    meta: {
      ...meta,
      site: cfg.site.baseUrl,
      overall,
      counts,
      pagesCrawled: site.pages.size,
      htmlPages: site.htmlPages().length,
      pagesRendered: rendered.available ? rendered.pages.size : 0,
      renderAvailable: rendered.available,
      renderSkipReason: rendered.reason,
      sitemapUrls: site.sitemap.urls.length,
      crawlTruncated: site.stoppedAtLimit,
    },
    checks: results.map((r) => ({
      id: r.id,
      title: r.title,
      status: r.status,
      durationMs: r.durationMs,
      error: r.error || null,
      counts: countBy(r.findings),
      // Distinct defects vs raw occurrences before collapsing.
      rawFindingCount: r.rawFindingCount ?? r.findings.length,
      findings: r.findings,
    })),
    regressions,
  };
}

function countBy(findings) {
  const out = { FAIL: 0, WARN: 0, INFO: 0 };
  for (const f of findings) out[f.severity] = (out[f.severity] || 0) + 1;
  return out;
}

export function renderConsole(report, cfg) {
  const lines = [];
  const bar = '='.repeat(72);
  lines.push(bar);
  lines.push(`  EMJAY WELLNESS QA REPORT — ${report.meta.overall}`);
  lines.push(`  ${report.meta.site}`);
  lines.push(`  ${report.meta.startedAt}`);
  lines.push(bar);
  lines.push('');
  lines.push(`  Pages crawled : ${report.meta.pagesCrawled} (${report.meta.htmlPages} HTML)`);
  lines.push(`  Pages rendered: ${report.meta.pagesRendered}${report.meta.renderAvailable ? '' : ` (skipped: ${report.meta.renderSkipReason})`}`);
  lines.push(`  Sitemap URLs  : ${report.meta.sitemapUrls}`);
  if (report.meta.crawlTruncated) lines.push(`  ! crawl hit the maxPages limit (${cfg.crawl.maxPages}) — coverage is partial`);
  lines.push('');
  lines.push(`  FAIL ${report.meta.counts.FAIL}   WARNING ${report.meta.counts.WARN}   INFO ${report.meta.counts.INFO}`);
  lines.push('');
  lines.push('-'.repeat(72));
  lines.push('  CHECK RESULTS');
  lines.push('-'.repeat(72));

  for (const check of report.checks) {
    const badge = LABEL[check.status].padEnd(7);
    const tally = check.counts.FAIL || check.counts.WARN || check.counts.INFO
      ? ` (${check.counts.FAIL} fail, ${check.counts.WARN} warn, ${check.counts.INFO} info)`
      : '';
    lines.push(`  ${badge} ${check.title}${tally}`);
    if (check.error) lines.push(`          ! check errored: ${check.error}`);

    const shown = check.findings
      .filter((f) => rank(f.severity) >= rank('WARN'))
      .sort((a, b) => rank(b.severity) - rank(a.severity))
      .slice(0, cfg.report.consoleFindingsPerCheck);
    for (const f of shown) {
      lines.push(`          [${LABEL[f.severity]}] ${f.title}`);
      if (f.url) lines.push(`                  ${f.url}`);
      if (f.detail) lines.push(`                  ${String(f.detail).split('\n').join('\n                  ')}`);
    }
    const hidden = check.findings.filter((f) => rank(f.severity) >= rank('WARN')).length - shown.length;
    if (hidden > 0) lines.push(`          ... and ${hidden} more (see the full report)`);
  }

  if (report.regressions.length) {
    lines.push('');
    lines.push('-'.repeat(72));
    lines.push('  KNOWN-DEFECT REGRESSIONS');
    lines.push('-'.repeat(72));
    for (const r of report.regressions) {
      lines.push(`  ${r.result.padEnd(12)} ${r.id}  ${r.title}`);
      if (r.result !== 'PASS' && r.detail) lines.push(`               ${r.detail}`);
    }
  }

  lines.push('');
  lines.push(bar);
  lines.push(`  OVERALL: ${report.meta.overall}`);
  lines.push(bar);
  return lines.join('\n');
}

export function renderMarkdown(report, cfg) {
  const md = [];
  md.push(`# Emjay Wellness — QA Report`);
  md.push('');
  md.push(`**Overall: ${report.meta.overall}**`);
  md.push('');
  md.push(`| | |`);
  md.push(`|---|---|`);
  md.push(`| Site | ${report.meta.site} |`);
  md.push(`| Run started | ${report.meta.startedAt} |`);
  md.push(`| Duration | ${(report.meta.durationMs / 1000).toFixed(1)}s |`);
  md.push(`| Pages crawled | ${report.meta.pagesCrawled} (${report.meta.htmlPages} HTML) |`);
  md.push(`| Pages rendered | ${report.meta.pagesRendered}${report.meta.renderAvailable ? '' : ' (browser pass skipped)'} |`);
  md.push(`| Sitemap URLs | ${report.meta.sitemapUrls} |`);
  md.push(`| FAIL / WARNING / INFO | ${report.meta.counts.FAIL} / ${report.meta.counts.WARN} / ${report.meta.counts.INFO} |`);
  md.push('');
  if (!report.meta.renderAvailable) {
    md.push(`> **Browser pass did not run.** ${report.meta.renderSkipReason}`);
    md.push(`> Colour, font, contrast, console, overflow and performance results are incomplete.`);
    md.push('');
  }
  if (report.meta.crawlTruncated) {
    md.push(`> **Crawl truncated** at the ${cfg.crawl.maxPages}-page limit. Coverage is partial.`);
    md.push('');
  }

  md.push('## Summary');
  md.push('');
  md.push('| Check | Status | Fail | Warning | Info |');
  md.push('|---|---|---:|---:|---:|');
  for (const c of report.checks) {
    md.push(`| ${c.title} | **${LABEL[c.status]}** | ${c.counts.FAIL} | ${c.counts.WARN} | ${c.counts.INFO} |`);
  }
  md.push('');

  if (report.regressions.length) {
    md.push('## Known-defect regressions');
    md.push('');
    md.push('| ID | Defect | Result | Detail |');
    md.push('|---|---|---|---|');
    for (const r of report.regressions) {
      md.push(`| ${r.id} | ${escapePipes(r.title)} | **${r.result}** | ${escapePipes(r.detail).slice(0, 200)} |`);
    }
    md.push('');
  }

  md.push('## Findings');
  md.push('');
  for (const severity of ORDER) {
    if (severity === 'PASS') continue;
    const group = report.checks.flatMap((c) => c.findings.filter((f) => f.severity === severity).map((f) => ({ ...f, check: c.title })));
    if (!group.length) continue;
    md.push(`### ${LABEL[severity]} (${group.length})`);
    md.push('');
    for (const f of group) {
      md.push(`- **${escapeMd(f.title)}** — _${f.check}_`);
      if (f.url) md.push(`  - \`${f.url}\``);
      if (f.detail) md.push(`  - ${escapeMd(String(f.detail)).replace(/\n\s*/g, '  \n    ')}`);
      if (f.fix) md.push(`  - Fix: ${escapeMd(f.fix)}`);
      if (f.defectId) md.push(`  - Regression id: \`${f.defectId}\``);
    }
    md.push('');
  }

  md.push('---');
  md.push('');
  md.push('_Generated by the Emjay QA suite. Read-only: no live-site content was modified._');
  return md.join('\n');
}

export function renderHtml(report) {
  const badge = (s) => `<span class="badge ${s.toLowerCase()}">${LABEL[s] || s}</span>`;
  const rows = report.checks
    .map(
      (c) => `<tr><td>${esc(c.title)}</td><td>${badge(c.status)}</td><td class="n">${c.counts.FAIL}</td><td class="n">${c.counts.WARN}</td><td class="n">${c.counts.INFO}</td></tr>`,
    )
    .join('\n');

  const regressionRows = report.regressions
    .map(
      (r) =>
        `<tr><td><code>${esc(r.id)}</code></td><td>${esc(r.title)}</td><td><span class="badge ${r.result === 'PASS' ? 'pass' : r.result === 'FAIL' ? 'fail' : 'warn'}">${r.result}</span></td><td>${esc(r.detail).slice(0, 240)}</td></tr>`,
    )
    .join('\n');

  const findingBlocks = ORDER.filter((s) => s !== 'PASS')
    .map((severity) => {
      const group = report.checks.flatMap((c) => c.findings.filter((f) => f.severity === severity).map((f) => ({ ...f, check: c.title })));
      if (!group.length) return '';
      return `<h3>${LABEL[severity]} (${group.length})</h3>` + group
        .map(
          (f) => `<div class="finding ${severity.toLowerCase()}">
  <div class="ft">${esc(f.title)} <span class="chk">${esc(f.check)}</span></div>
  ${f.url ? `<div class="fu"><a href="${esc(f.url)}" rel="noopener nofollow">${esc(f.url)}</a></div>` : ''}
  ${f.detail ? `<pre class="fd">${esc(String(f.detail))}</pre>` : ''}
  ${f.fix ? `<div class="fx">Fix: ${esc(f.fix)}</div>` : ''}
</div>`,
        )
        .join('\n');
    })
    .join('\n');

  return `<!doctype html>
<html lang="en-AU"><head><meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Emjay Wellness QA Report</title>
<style>
  :root { --bg:#fcfcfb; --ink:#2e2c27; --soft:#6b6a63; --line:#e4e3dc; --fail:#b3261e; --warn:#a8620a; --info:#4a6572; --pass:#2e6b3f; }
  * { box-sizing: border-box; }
  body { margin:0; padding:2rem 1.25rem 4rem; background:var(--bg); color:var(--ink);
         font:15px/1.55 -apple-system, "Segoe UI", sans-serif; }
  main { max-width: 980px; margin: 0 auto; }
  h1 { font-size:1.7rem; margin:0 0 .25rem; }
  h2 { font-size:1.2rem; margin:2.5rem 0 .75rem; border-bottom:1px solid var(--line); padding-bottom:.35rem; }
  h3 { font-size:1rem; margin:1.75rem 0 .5rem; color:var(--soft); text-transform:uppercase; letter-spacing:.06em; }
  .sub { color:var(--soft); margin:0 0 1.5rem; }
  .overall { display:inline-block; padding:.4rem .9rem; border-radius:6px; font-weight:600; color:#fff; }
  .overall.fail { background:var(--fail); } .overall.warning { background:var(--warn); } .overall.pass { background:var(--pass); }
  table { width:100%; border-collapse:collapse; font-size:.92rem; }
  th, td { text-align:left; padding:.5rem .6rem; border-bottom:1px solid var(--line); vertical-align:top; }
  th { color:var(--soft); font-weight:600; }
  td.n { text-align:right; font-variant-numeric:tabular-nums; }
  .badge { display:inline-block; padding:.1rem .5rem; border-radius:4px; font-size:.78rem; font-weight:600; color:#fff; }
  .badge.fail { background:var(--fail); } .badge.warn, .badge.warning { background:var(--warn); }
  .badge.info { background:var(--info); } .badge.pass { background:var(--pass); }
  .finding { border-left:3px solid var(--line); padding:.5rem 0 .5rem .8rem; margin:.6rem 0; }
  .finding.fail { border-left-color:var(--fail); } .finding.warn { border-left-color:var(--warn); }
  .finding.info { border-left-color:var(--info); }
  .ft { font-weight:600; }
  .chk { font-weight:400; color:var(--soft); font-size:.82rem; margin-left:.4rem; }
  .fu a { color:var(--info); font-size:.85rem; word-break:break-all; }
  .fd { background:#f4f3ef; padding:.5rem .6rem; border-radius:4px; font-size:.82rem; overflow-x:auto; white-space:pre-wrap; margin:.35rem 0; }
  .fx { font-size:.85rem; color:var(--soft); }
  .wrap { overflow-x:auto; }
  footer { margin-top:3rem; color:var(--soft); font-size:.85rem; border-top:1px solid var(--line); padding-top:1rem; }
</style></head><body><main>
<h1>Emjay Wellness — QA Report</h1>
<p class="sub">${esc(report.meta.site)} · ${esc(report.meta.startedAt)}</p>
<p><span class="overall ${report.meta.overall.toLowerCase()}">OVERALL: ${report.meta.overall}</span></p>
<p class="sub">${report.meta.pagesCrawled} pages crawled · ${report.meta.pagesRendered} rendered · ${report.meta.counts.FAIL} fail · ${report.meta.counts.WARN} warning · ${report.meta.counts.INFO} info</p>
${report.meta.renderAvailable ? '' : `<p class="finding warn"><strong>Browser pass did not run.</strong> ${esc(report.meta.renderSkipReason || '')}</p>`}
<h2>Check summary</h2>
<div class="wrap"><table><thead><tr><th>Check</th><th>Status</th><th class="n">Fail</th><th class="n">Warn</th><th class="n">Info</th></tr></thead><tbody>
${rows}
</tbody></table></div>
${report.regressions.length ? `<h2>Known-defect regressions</h2><div class="wrap"><table><thead><tr><th>ID</th><th>Defect</th><th>Result</th><th>Detail</th></tr></thead><tbody>${regressionRows}</tbody></table></div>` : ''}
<h2>Findings</h2>
${findingBlocks || '<p>No findings.</p>'}
<footer>Generated by the Emjay QA suite. Read-only: no live-site content was modified.</footer>
</main></body></html>`;
}

export async function writeReports(report, cfg) {
  const outDir = path.resolve(cfg.report.outDir);
  await mkdir(outDir, { recursive: true });
  const written = [];

  if (cfg.report.formats.includes('json')) {
    const p = path.join(outDir, 'qa-report.json');
    await writeFile(p, JSON.stringify(report, null, 2));
    written.push(p);
  }
  if (cfg.report.formats.includes('md')) {
    const p = path.join(outDir, 'qa-report.md');
    await writeFile(p, renderMarkdown(report, cfg));
    written.push(p);
  }
  if (cfg.report.formats.includes('html')) {
    const p = path.join(outDir, 'qa-report.html');
    await writeFile(p, renderHtml(report));
    written.push(p);
  }
  return written;
}

function esc(s) {
  return String(s ?? '').replace(/[&<>"]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;' }[c]));
}
function escapeMd(s) {
  return String(s ?? '').replace(/([*_`])/g, '\\$1');
}
function escapePipes(s) {
  return String(s ?? '').replace(/\|/g, '\\|').replace(/\n/g, ' ');
}

File: qa/emjay_qa/report.py
"""Report rendering: console summary, Markdown, JSON and a standalone HTML
page.

The report states what was measured and where. It contains no rewritten copy,
no SEO suggestions and no design direction.
"""
from __future__ import annotations

import html as _html
import json
from collections import defaultdict
from typing import Dict, List

from .models import Severity
from .runner import CheckResult, RunResult

MAX_FINDINGS_PER_CHECK = 25

BADGE = {
    Severity.PASS: "PASS",
    Severity.INFO: "INFO",
    Severity.WARNING: "WARNING",
    Severity.FAIL: "FAIL",
}


def _by_category(run: RunResult) -> Dict[str, List[CheckResult]]:
    grouped: Dict[str, List[CheckResult]] = defaultdict(list)
    for result in run.results:
        grouped[result.meta.category].append(result)
    return dict(sorted(grouped.items()))


def console_summary(run: RunResult) -> str:
    summary = run.summary
    lines = [
        "",
        f"VERDICT: {run.verdict.value}",
        f"  checks   {summary['checks_passed']} pass / {summary['checks_warning']} warning / "
        f"{summary['checks_failed']} fail / {summary['checks_info']} info",
        f"  findings {summary['findings_fail']} fail / {summary['findings_warning']} warning / "
        f"{summary['findings_info']} info",
        f"  coverage {summary['pages_crawled']} pages crawled, {summary['pages_rendered']} rendered in a browser",
    ]
    if summary["checks_skipped"]:
        lines.append(f"  skipped  {summary['checks_skipped']} check(s) did not run")
    failures = [r for r in run.results if r.status is Severity.FAIL]
    if failures:
        lines.append("")
        lines.append("  failing checks:")
        for result in failures:
            lines.append(f"    - {result.meta.id}: {len(result.findings)} finding(s)")
    return "\n".join(lines) + "\n"


def to_json(run: RunResult) -> str:
    payload = {
        "site": run.site.config.base_url,
        "verdict": run.verdict.value,
        "crawl_started_at": run.site.started_at,
        "crawl_finished_at": run.site.finished_at,
        "analysis_started_at": run.started_at,
        "analysis_finished_at": run.finished_at,
        "browser_used": run.site.browser_used,
        "summary": run.summary,
        "notes": run.site.notes,
        "checks": [
            {
                "id": result.meta.id,
                "name": result.meta.name,
                "category": result.meta.category,
                "status": result.status.value,
                "skipped_reason": result.skipped_reason,
                "error": result.error,
                "counts": result.counts,
                "findings": [f.to_dict() for f in result.findings],
            }
            for result in run.results
        ],
    }
    return json.dumps(payload, indent=2, ensure_ascii=False)


def _finding_lines(result: CheckResult, severity: Severity) -> List[str]:
    lines = []
    items = [f for f in result.findings if f.severity is severity]
    for item in items[:MAX_FINDINGS_PER_CHECK]:
        location = f" `{item.url}`" if item.url else ""
        lines.append(f"- **{item.title}**{location}")
        if item.detail:
            lines.append(f"  - {item.detail}")
    if len(items) > MAX_FINDINGS_PER_CHECK:
        lines.append(f"- _...and {len(items) - MAX_FINDINGS_PER_CHECK} more (see the JSON report)_")
    return lines


def to_markdown(run: RunResult) -> str:
    site = run.site
    summary = run.summary
    out: List[str] = []
    out.append("# Emjay Wellness: automated QA report")
    out.append("")
    out.append(f"**Verdict: {run.verdict.value}**")
    out.append("")
    out.append(f"- Site: {site.config.base_url}")
    out.append(f"- Crawl: {site.started_at} to {site.finished_at} (UTC)")
    out.append(f"- Pages crawled: {summary['pages_crawled']}, rendered in a browser: {summary['pages_rendered']}")
    out.append(f"- Checks: {summary['checks_passed']} PASS, {summary['checks_warning']} WARNING, "
               f"{summary['checks_failed']} FAIL, {summary['checks_info']} INFO"
               + (f", {summary['checks_skipped']} not run" if summary["checks_skipped"] else ""))
    out.append(f"- Findings: {summary['findings_fail']} FAIL, {summary['findings_warning']} WARNING, "
               f"{summary['findings_info']} INFO")
    out.append("- Method: read-only GET/HEAD requests plus headless rendering. No form was submitted and nothing "
               "on the live site was changed.")
    if not summary["pages_crawled"] or not any(p.fetch and p.fetch.ok for p in site.pages.values()):
        out.append("")
        out.append("> **The site could not be reached during this run, so the result below is a verdict on the "
                   "run, not on the site.** Re-run from a network that can reach it.")
    if site.notes:
        out.append("")
        out.append("**Run notes**")
        for note in site.notes:
            out.append(f"- {note}")

    out.append("")
    out.append("## Check status by area")
    out.append("")
    out.append("| Area | Check | Status | Findings |")
    out.append("| --- | --- | --- | --- |")
    for category, results in _by_category(run).items():
        for result in sorted(results, key=lambda r: (-r.status.rank, r.meta.id)):
            counts = result.counts
            detail = ", ".join(
                f"{counts[level]} {level.lower()}"
                for level in ("FAIL", "WARNING", "INFO")
                if counts[level]
            ) or "none"
            out.append(f"| {category} | `{result.meta.id}` | {BADGE[result.status]} | {detail} |")

    for severity, heading in (
        (Severity.FAIL, "## FAIL: must be fixed before Emjay is declared GREEN"),
        (Severity.WARNING, "## WARNING: review and decide"),
        (Severity.INFO, "## INFO: discovery and coverage notes"),
    ):
        relevant = [r for r in run.results if any(f.severity is severity for f in r.findings)]
        if not relevant:
            continue
        out.append("")
        out.append(heading)
        for result in sorted(relevant, key=lambda r: r.meta.id):
            out.append("")
            out.append(f"### `{result.meta.id}` {result.meta.name}")
            if result.meta.description:
                out.append(f"_{result.meta.description.splitlines()[0]}_")
            out.extend(_finding_lines(result, severity))

    out.append("")
    out.append("## Known-defect regression status")
    out.append("")
    defect_findings = [f for f in run.findings if f.check_id == "regression.known_defect"]
    if not defect_findings:
        out.append("All registry entries passed.")
    else:
        out.append("| Defect | Status in registry | Result |")
        out.append("| --- | --- | --- |")
        for item in defect_findings:
            defect_id = item.evidence.get("defect_id", "-")
            status = item.evidence.get("status", "-")
            out.append(f"| {defect_id} | {status} | {BADGE[item.severity]}: {item.title} |")

    out.append("")
    out.append("## Scope and limits")
    out.append("")
    out.append("- Contrast is measured only where the effective background is a solid computed colour. Text over "
               "images, video or gradients is skipped rather than guessed.")
    out.append("- Performance figures are indicative signals from the crawl and browser navigation timings, not a "
               "Lighthouse audit.")
    out.append("- Brand palette and font enforcement stay in discovery mode until `brand.approved_colours` and "
               "`brand.approved_fonts` are populated in `qa/config/site.yaml`.")
    out.append("- Forms are inspected structurally. Nothing is ever submitted.")
    out.append("")
    return "\n".join(out)


def to_html(run: RunResult) -> str:
    summary = run.summary
    verdict_colour = {"PASS": "#1B7F4B", "WARNING": "#9A6400", "FAIL": "#B3261E", "INFO": "#3B5566"}[run.verdict.value]

    def esc(value: str) -> str:
        return _html.escape(str(value or ""))

    rows = []
    for category, results in _by_category(run).items():
        for result in sorted(results, key=lambda r: (-r.status.rank, r.meta.id)):
            counts = result.counts
            detail = ", ".join(
                f"{counts[level]} {level.lower()}" for level in ("FAIL", "WARNING", "INFO") if counts[level]
            ) or "none"
            rows.append(
                f"<tr><td>{esc(category)}</td><td><code>{esc(result.meta.id)}</code></td>"
                f"<td class='s-{result.status.value.lower()}'>{result.status.value}</td><td>{esc(detail)}</td></tr>"
            )

    sections = []
    for severity, heading in (
        (Severity.FAIL, "FAIL"),
        (Severity.WARNING, "WARNING"),
        (Severity.INFO, "INFO"),
    ):
        blocks = []
        for result in sorted(run.results, key=lambda r: r.meta.id):
            items = [f for f in result.findings if f.severity is severity]
            if not items:
                continue
            entries = "".join(
                f"<li><strong>{esc(item.title)}</strong>"
                + (f" <code>{esc(item.url)}</code>" if item.url else "")
                + (f"<div class='detail'>{esc(item.detail)}</div>" if item.detail else "")
                + "</li>"
                for item in items[:MAX_FINDINGS_PER_CHECK]
            )
            more = (
                f"<li class='more'>...and {len(items) - MAX_FINDINGS_PER_CHECK} more in the JSON report</li>"
                if len(items) > MAX_FINDINGS_PER_CHECK
                else ""
            )
            blocks.append(
                f"<section><h3><code>{esc(result.meta.id)}</code> {esc(result.meta.name)}</h3>"
                f"<ul>{entries}{more}</ul></section>"
            )
        if blocks:
            sections.append(f"<h2 class='s-{severity.value.lower()}'>{heading}</h2>" + "".join(blocks))

    unreachable = not any(p.fetch and p.fetch.ok for p in run.site.pages.values())
    unreachable_banner = (
        '<p class="s-fail">The site could not be reached during this run, so this is a verdict on the run, '
        "not on the site.</p>"
        if unreachable
        else ""
    )
    return f"""<title>Emjay QA Report</title>
<style>
  :root {{ --bg:#FAF9F5; --fg:#1F2421; --muted:#5C6660; --line:#DFDCD2; --card:#FFFFFF; }}
  @media (prefers-color-scheme: dark) {{
    :root:not([data-theme="light"]) {{ --bg:#15181A; --fg:#ECEAE3; --muted:#9AA5A0; --line:#2C3134; --card:#1D2124; }}
  }}
  :root[data-theme="dark"] {{ --bg:#15181A; --fg:#ECEAE3; --muted:#9AA5A0; --line:#2C3134; --card:#1D2124; }}
  body {{ background:var(--bg); color:var(--fg); font:16px/1.55 -apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif;
         margin:0; padding:2rem 1.25rem 4rem; }}
  main {{ max-width:1080px; margin:0 auto; }}
  h1 {{ font-size:1.7rem; margin:0 0 .25rem; }}
  h2 {{ margin-top:2.5rem; border-bottom:1px solid var(--line); padding-bottom:.35rem; }}
  h3 {{ font-size:1rem; margin:1.4rem 0 .4rem; }}
  .verdict {{ display:inline-block; padding:.35rem .9rem; border-radius:999px; color:#fff;
              background:{verdict_colour}; font-weight:600; letter-spacing:.02em; }}
  .meta {{ color:var(--muted); font-size:.92rem; }}
  table {{ width:100%; border-collapse:collapse; margin-top:1rem; font-size:.9rem; }}
  th,td {{ text-align:left; padding:.45rem .6rem; border-bottom:1px solid var(--line); vertical-align:top; }}
  th {{ color:var(--muted); font-weight:600; }}
  .table-wrap {{ overflow-x:auto; }}
  code {{ font-family:ui-monospace,SFMono-Regular,Menlo,monospace; font-size:.85em; }}
  ul {{ padding-left:1.1rem; }}
  li {{ margin:.35rem 0; }}
  .detail {{ color:var(--muted); font-size:.88rem; word-break:break-word; }}
  .s-fail {{ color:#B3261E; font-weight:600; }}
  .s-warning {{ color:#9A6400; font-weight:600; }}
  .s-pass {{ color:#1B7F4B; font-weight:600; }}
  .s-info {{ color:#3B5566; font-weight:600; }}
  section {{ background:var(--card); border:1px solid var(--line); border-radius:10px; padding:.6rem 1rem; margin:.8rem 0; }}
</style>
<main>
  <h1>Emjay Wellness: automated QA report</h1>
  <p><span class="verdict">{run.verdict.value}</span></p>
  <p class="meta">
    {esc(run.site.config.base_url)} &middot; crawl {esc(run.site.started_at)} to {esc(run.site.finished_at)} UTC<br>
    {summary['pages_crawled']} pages crawled, {summary['pages_rendered']} rendered in a browser &middot;
    {summary['checks_passed']} PASS / {summary['checks_warning']} WARNING / {summary['checks_failed']} FAIL /
    {summary['checks_info']} INFO<br>
    Read-only GET and HEAD requests plus headless rendering. No form submitted, nothing on the live site changed.
  </p>
  {unreachable_banner}
  <h2>Check status by area</h2>
  <div class="table-wrap">
  <table><thead><tr><th>Area</th><th>Check</th><th>Status</th><th>Findings</th></tr></thead>
  <tbody>{''.join(rows)}</tbody></table>
  </div>
  {''.join(sections)}
</main>
"""
